memory.c

```diff
@@ -0,0 +1,329 @@
+    for (int vpn = kernel_text_start; vpn < kernel_brk_start; vpn++) {  //
+        pte_t *entry = (pte_t *)(p_address_region0_base + vpn * sizeof(pte_t));
+        entry->valid = 1;
+        if (vpn < kernel_data_start) { // Set protection bits appropriately
+            entry->prot = PROT_READ | PROT_EXEC;  // Text section: read+execute
+        } else {
+            entry->prot = PROT_READ | PROT_WRITE;  // Data/heap: read+write
+        }
+        entry->pfn = vpn;
+        frame_bitMap[entry->pfn] = 1;  // Mark the *physical frame* as used in the bitmap
+    }
+    TracePrintf(0, "Kernel text, data, and heap initialized with %d total entries\n", kernel_brk_start);
+
+    void * kernel_brk = UP_TO_PAGE(kernel_brk_start * PAGESIZE);  // Align to page boundary
+    if (SetKernelBrk(kernel_brk) != 0) {
+        TracePrintf(0, "ERROR: Failed to set kernel break to address %p\n", kernel_brk);
+        return ERROR;  // Out of physical memory
+    }
+
+    // Initialize the kernel stack in the page table
+    int base_kernelStack_page = KERNEL_STACK_BASE / PAGESIZE;  // Calculate the base page number for kernel stack
+    int kernelStack_limit_page = KERNEL_STACK_LIMIT / PAGESIZE;  // Calculate the limit page number for kernel stack
+    for (int vpn = base_kernelStack_page; vpn < kernelStack_limit_page; vpn ++) {
+        pte_t *entry = (pte_t *)(p_address_region0_base + vpn * sizeof(pte_t));
+        entry->valid = 1;
+        entry->prot = PROT_READ | PROT_WRITE;
+        entry->pfn = vpn;       // Set frame numbers to the ones proceeding kernel text, data, and heap frames
+        frame_bitMap[vpn] = 1;  // Mark the frame as used in the bitmap
+    }
+    TracePrintf(0, "Kernel stack initialized with %d total entries\n", (KERNEL_STACK_LIMIT - KERNEL_STACK_BASE) / sizeof(pte_t));
+
+    // Set page table registers to tell hardware where to find the page table
+    int numPages_inRegion0 = VMEM_0_LIMIT / PAGESIZE;  // Number of pages in Region 0
+    WriteRegister(REG_PTBR0, p_address_region0_base); // Set base address of region0 page table in hardware
+    WriteRegister(REG_PTLR0, numPages_inRegion0); // Set limit of region0 page table in hardware
+    TracePrintf(0, "Wrote region 0 page table base register to %d\n", p_address_region0_base);
+    TracePrintf(0, "Wrote region 0 page table limit register to %d\n", numPages_inRegion0);
+
+    // Set up the base physicall adress of region 1
+    unsigned int p_address_region1_base = p_address_region0_base + numPages_inRegion0 * sizeof(pte_t); // Place region1 PT at the end of region0 PT
+    int vpn = VMEM_1_BASE / PAGESIZE; // Calculate the virtual page number for the first page of Region 1
+    pte_t *entry = (pte_t *)(p_address_region1_base);
+    entry->valid = 1;
+    entry->prot = PROT_READ | PROT_WRITE;
+    entry->pfn = 1; // Set frame number to 1 since it is the first page of region 1
+    frame_bitMap[vpn] = 1;  // Mark the frame as used in the bitmap

```

Comment:
why are you making a single page at the bottom of region 1 valid? this doesn't make sense to me. the stack would be at the top. and any pages for the data or text will be setup in loadprogram.